### PR TITLE
Pytorch Optimizations

### LINKED ISSUE #1990: Implement some easy PyTorch optimizations
See https://twitter.com/karpathy/status/1299921324333170689?s=03

### PR DESCRIPTION
**Pull request by mugoh**
Implements optimizations for torch.
resolves #1990

### CODE REVIEW

**Comment by mugoh** [comment]
An optimization strategy mentioned in [the resource](https://twitter.com/karpathy/status/1299921324333170689?s=03) linked to the issue (#1990) suggests replacing `optim.zero_grad()` during training with setting the gradient of the parameters to `None`. Something similar to:
```python
for group in optim.param_groups:
    for param in group['params']:
        param.grad = None
```

This option has been added to Pytorch 1.7 as an optional argument `set_to_none`, used as `optim.zero_grad(set_to_none=False)`, but with a [warning](https://github.com/pytorch/pytorch/blob/76ee58e2ec48d6922c49d647b276c1575a96aa09/torch/optim/optimizer.py#L167-L186) on setting grads to `None`.

So I'm requesting some clarity if setting grads to None should be added for optimization.

**Comment by krzentner** [comment]
Hi Mugoh,

Sorry for taking so long to respond to your PR. We would appreciate our pytorch branch getting faster.
About the `torch.as_tensor` optimizations, can you just modify [`np_to_torch`](https://github.com/rlworkgroup/garage/blob/master/src/garage/torch/_functions.py#L137) instead, and replace the other lines with that function?

The CNN optimization looks useful, since our inputs are always the same size (at least within the sampling process).

About the setting grad to `None` optimization, it looks useful, but we would prefer not to require a specific version of torch.
Can you add a function for it in `garage.torch._functions`?

Thank you for helping to improve garage.

**Comment by mugoh** [comment]
Thanks for the review krzentner. Sure, I'll make these changes and update the PR. 

For setting grads to `None`, after the changes, I think I can later update the existing algorithms to use it, and get a review on where it shouldn't work.

**Comment by mugoh** [comment]
@krzentner, I've made the changes you suggested. About the setting grads to `None` optimization, I added a function for that -- are there specific PyTorch algorithms that should work with this tweak, and should I go ahead and alter them in this PR?

**Comment by krzentner** [comment]
This looks good. I think all of the single-task off-policy pytorch algorithms (DQN, DDPG, TD3, and SAC) should benefit from this. The on-policy algorithms (VPG, PPO, TRPO) should work with it too, but I'm not 100% sure it won't mess up MAML. You could do that in this PR or another PR.

By the way, there were a few formatting issues that made this change not pass pre-commit. I've pushed up a branch that fixes them [to a branch named kr-optimize-torch](https://github.com/rlworkgroup/garage/tree/kr-optimize-torch). If another reviewer approves this PR I'm happy to merge it.

File: src/garage/torch/algos/maml.py
Comment (yeukfu) [inline src/garage/torch/algos/maml.py:169]: Just curious why we need to use `obs.astype(np.float32)` instead of just `obs`.

**Review by yeukfu — APPROVED** [review]

**Review by krzentner — APPROVED** [review]

File: src/garage/torch/algos/maml.py
Comment (mugoh) [reply src/garage/torch/algos/maml.py:169]: Oh I missed that. Resolving

**Comment by mugoh** [comment]
> This looks good. I think all of the single-task off-policy pytorch algorithms (DQN, DDPG, TD3, and SAC) should benefit from this. The on-policy algorithms (VPG, PPO, TRPO) should work with it too, but I'm not 100% sure it won't mess up MAML. You could do that in this PR or another PR.
> 
> By the way, there were a few formatting issues that made this change not pass pre-commit. I've pushed up a branch that fixes them [to a branch named kr-optimize-torch](https://github.com/rlworkgroup/garage/tree/kr-optimize-torch). If another reviewer approves this PR I'm happy to merge it.

Alright. I'll implement that for both the mentioned onpolicy and offpolicy algorithms. For MAML, I'll check if the change brings significant deviation in performance.

**Comment by mugoh** [comment]
Here is the performance of MAML with and without the change that sets grads to `None` instead of calling `zero_grad`. I just took snapshots of the tensorboard logs. This is the actual script I used (taken from the docs) and modified to correct errors:

```python
#!/usr/bin/env python3
"""This is an example to train MAML-VPG on HalfCheetahDirEnv environment."""
# pylint: disable=no-value-for-parameter
import click
import torch

from garage import wrap_experiment
from garage.envs import GymEnv, normalize
from garage.envs.mujoco import HalfCheetahDirEnv
from garage.experiment import MetaEvaluator
from garage.experiment.deterministic import set_seed
from garage.experiment.task_sampler import SetTaskSampler
from garage.sampler import RaySampler
from garage.torch.algos import MAMLPPO
from garage.torch.policies import GaussianMLPPolicy
from garage.torch.value_functions import GaussianMLPValueFunction
from garage.trainer import Trainer


@click.command()
@click.option('--seed', default=1)
@click.option('--epochs', default=300)
@click.option('--episodes_per_task', default=40)
@click.option('--meta_batch_size', default=20)
@wrap_experiment(snapshot_mode='all')
def maml_ppo_half_cheetah_dir(ctxt, seed, epochs, episodes_per_task,
                              meta_batch_size):
    """Set up environment and algorithm and run the task.
    """
    set_seed(seed)
    max_episode_length = 100
    env = normalize(GymEnv(HalfCheetahDirEnv(),
                           max_episode_length=max_episode_length),
                    expected_action_scale=10.)

    policy = GaussianMLPPolicy(
        env_spec=env.spec,
        hidden_sizes=(64, 64),
        hidden_nonlinearity=torch.tanh,
        output_nonlinearity=None,
    )

    value_function = GaussianMLPValueFunction(env_spec=env.spec,
                                              hidden_sizes=(32, 32),
                                              hidden_nonlinearity=torch.tanh,
                                              output_nonlinearity=None)

    task_sampler = SetTaskSampler(
        HalfCheetahDirEnv,
        wrapper=lambda env, _: normalize(GymEnv(
            env, max_episode_length=max_episode_length),
            expected_action_scale=10.))

    meta_evaluator = MetaEvaluator(test_task_sampler=task_sampler,
                                   n_test_tasks=2,
                                   n_test_episodes=10)

    trainer = Trainer(ctxt)

    algo = MAMLPPO(env=env,
                   policy=policy,
                   # sampler=sampler,
                   task_sampler=task_sampler,
                   value_function=value_function,
                   meta_batch_size=meta_batch_size,
                   discount=0.99,
                   gae_lambda=1.,
                   inner_lr=0.1,
                   num_grad_updates=1,
                   meta_evaluator=meta_evaluator)

    trainer.setup(algo, env)
    trainer.train(n_epochs=epochs,
                  batch_size=episodes_per_task * env.spec.max_episode_length)


maml_ppo_half_cheetah_dir()

```

***

The results are organised as the logger grouped the logs:

- Setting grads to `None`: Green plots
- Calling `optimizer.zero_grad`: White plots

**1. Averages**

Average-Screenshot from 2020-12-10 20-01-15


**2. MetaTest**

Metatest-Screenshot from 2020-12-10 20-01-48


**3. Backward**

Backward-Screenshot from 2020-12-10 20-02-28



**4. Forward**

Forward-Screenshot from 2020-12-10 20-02-49

**Comment by ryanjulian** [comment]
Wow, great work @mugoh ! Since the premise of this is performance improvement, can you give us some info on the wall-clock time for the optimized vs non-optimized experiments? You can use the "WALL" option on TensorBoard (or you can send us a link to your results using tensorboard.dev).

**Comment by mugoh** [comment]
> Wow, great work @mugoh ! Since the premise of this is performance improvement, can you give us some info on the wall-clock time for the optimized vs non-optimized experiments? You can use the "WALL" option on TensorBoard (or you can send us a link to your results using tensorboard.dev).

Thanks @ryanjulian! I wish I had known of tensorboard.dev earlier. I've only run MAML since I wished to confirm the optimizations don't break it's performance. I'll run some of the other algorithms and share the results.

**Comment by ryanjulian** [comment]
Just MAML with and without the new optimization should be enough! But by all means, the more the merrier.

**Review by irisliucy — APPROVED** [review]

**Comment by ryanjulian** [comment]
@mugoh No rush, just checking in to see if you found any wall clock time improvements from this optimization.

**Comment by mugoh** [comment]
As of now, not yet - I haven't found significant wall clock time improvements for short runs. I'll test on longer ones and share the result. Thanks for the reminder @ryanjulian

**Comment by mugoh** [comment]
Tensorboard Dev experiment links

- [TD3](https://tensorboard.dev/experiment/ajvmMnSgRpWHoS1d5Jk2sg/)


_Will update with more runs_

**Comment by ryanjulian** [comment]
Looks great, thank you so much! Let's work on getting this merged ASAP

**Comment by mugoh** [comment]
> Looks great, thank you so much! Let's work on getting this merged ASAP

Great! There are some runtime errors I introduced in the process. Working on fixing all of them

**Comment by mugoh** [comment]
Was any further addition needed on this @ryanjulian ?

**Comment by ryanjulian** [comment]
I defer to @haydenshively and @gitanshu who have been paying attention to PRs more closely.

**Comment by haydenshively** [comment]
@mugoh Looks good to me! Thanks for all your work on this, we really appreciate it

**Comment by haydenshively** [comment]
@Mergifyio rebase

**Comment by haydenshively** [comment]
@mugoh I just ran garage's CI on this, and it looks like you're not passing pre-commit. Could you try to fix that before we merge this? More info [here](https://github.com/rlworkgroup/garage/blob/56889ac141819a4c57588df80975b725a03d758b/CONTRIBUTING.md#preparing-your-repo-to-make-contributions). Thanks!

**Comment by mugoh** [comment]
> @mugoh I just ran garage's CI on this, and it looks like you're not passing pre-commit. Could you try to fix that before we merge this? More info [here](https://github.com/rlworkgroup/garage/blob/56889ac141819a4c57588df80975b725a03d758b/CONTRIBUTING.md#preparing-your-repo-to-make-contributions). Thanks!

Working on the fix..

**Comment by mugoh** [comment]
@haydenshively I have been having a flake8 exception on the pre-commit and still can't resolve it. The complete log is here: #2258

**Comment by mugoh** [comment]
I've resolved the pre-commit errors @haydenshively

**Comment by mugoh** [comment]
Fixed the initial errors. Have a look at this again @haydenshively

**Comment by haydenshively** [comment]
Hi @mugoh,

Thanks for resolving those errors. I'm going to run our CI again to make sure everything passes. In the meantime, it looks like you removed the executable permission from a number of our `.sh` and `.py` example scripts (see changes in your most recent commit, 9a16cc3). Could you undo those changes please? [These instructions](https://gist.github.com/liitfr/86c77bf0ea4a627ab9b62ad30c32c251) may help. Thanks!

**Comment by mugoh** [comment]
Hi @haydenshively, I undid the permission changes

**Comment by haydenshively** [comment]
@mugoh thanks for all your work on this. It looks like [the CI](https://github.com/rlworkgroup/garage/runs/2224892367?check_suite_focus=true) is now being picky about the capitalization of your commit messages, but we don't want to make you (and future contributors) deal with stuff like that.

We've decided to remove flake8 from garage to make passing pre-commit easier. We plan to remove it by the end of the week, at which point you should be able to rebase your PR on master and get it merged more easily.

In the meantime, if you do want to attend to flake8's pesky warnings and get this merged sooner, you're more than welcome to.

**Comment by mugoh** [comment]
> @mugoh thanks for all your work on this. It looks like [the CI](https://github.com/rlworkgroup/garage/runs/2224892367?check_suite_focus=true) is now being picky about the capitalization of your commit messages, but we don't want to make you (and future contributors) deal with stuff like that.
> 
> We've decided to remove flake8 from garage to make passing pre-commit easier. We plan to remove it by the end of the week, at which point you should be able to rebase your PR on master and get it merged more easily.
> 
> In the meantime, if you do want to attend to flake8's pesky warnings and get this merged sooner, you're more than welcome to.

Thanks @haydenshively. I'll wait till the end of the week when flake8 is removed.

**Comment by mugoh** [comment]
We could check this again @haydenshively. I tried to attend to the flake8 warnings

**Comment by haydenshively** [comment]
Thanks @mugoh, and sorry for the confusion about standards. I'll start the tests.

**Comment by haydenshively** [comment]
@krzentner Tests have passed on this. Please force merge when you get a chance

**Comment by krzentner** [comment]
Sorry this took so long for me to see. I rebased / squashed it. It should auto-merge now.

**Pull request merged**
